File: domain/src/main/java/com/example/app/movie/dto/response/MovieResponse.java
package com.example.app.movie.dto.response;

import com.example.app.movie.domain.Movie;
import com.example.app.movie.domain.Showtime;
import com.example.app.movie.domain.Theater;
import lombok.Builder;

import java.time.LocalDate;
import java.time.format.DateTimeFormatter;
import java.util.Comparator;
import java.util.List;

@Builder
public record MovieResponse(
        Long id,
        String title,
        String description,
        String status,
        String rating,
        String genre,
        String thumbnail,
        int runningTime,
        LocalDate releaseDate,
        List<String> showtimes,
        List<String> theaters
){
    public static MovieResponse toResponse(Movie movie) {
        var showtimes = movie.showtimes()
                .stream()
                .sorted(Comparator.comparing(Showtime::start))
                .map(showtime -> String.format("%s ~ %s",
                        showtime.start().format(DateTimeFormatter.ofPattern("HH:mm")),
                        showtime.end().format(DateTimeFormatter.ofPattern("HH:mm"))))
                .toList();

        var theaters = movie.theaters()
                .stream()
                .map(Theater::name)
                .toList();

        return MovieResponse.builder()
                .id(movie.id())
                .title(movie.title())
                .description(movie.description())
                .status(movie.status().getDescription())
                .rating(movie.rating().getDescription())
                .genre(movie.genre().getDescription())
                .thumbnail(movie.thumbnail())
                .runningTime(movie.runningTime())
                .releaseDate(movie.releaseDate())
                .showtimes(showtimes)
                .theaters(theaters)
                .build();
    }
}

File: movie/api/src/main/kotlin/com/example/movie/dto/MovieResponse.kt
package com.example.movie.dto

import com.example.movie.domain.movie.model.Genre
import com.example.movie.domain.movie.model.Movie
import com.example.movie.domain.movie.model.Rating
import com.example.movie.domain.screening.model.Screening
import com.example.movie.domain.theater.model.Theater
import java.time.LocalDate
import java.time.LocalDateTime

data class MovieResponse(
    val id: Long,
    val title: String,
    val rating: Rating,
    val genre: GenreResponse,
    val releaseDate: LocalDate,
    val thumbnailUrl: String,
    val runningTime: Int,
    val screenings: List<ScreeningResponse>
) {
    companion object {
        fun from(movie: Movie, screenings: List<Screening>): MovieResponse {
            return MovieResponse(
                id = movie.id,
                title = movie.title,
                rating = movie.rating,
                genre = GenreResponse.from(movie.genre),
                releaseDate = movie.releaseDate,
                thumbnailUrl = movie.thumbnailUrl,
                runningTime = movie.runningTime,
                screenings = screenings.map { ScreeningResponse.from(it) }
            )
        }
    }
}

data class GenreResponse(
    val id: Long,
    val name: String
) {
    companion object {
        fun from(genre: Genre): GenreResponse {
            return GenreResponse(
                id = genre.id,
                name = genre.name
            )
        }
    }
}

data class ScreeningResponse(
    val id: Long,
    val screeningTime: LocalDateTime,
    val screeningEndTime: LocalDateTime,
    val theater: TheaterResponse
) {
    companion object {
        fun from(screening: Screening): ScreeningResponse {
            return ScreeningResponse(
                id = screening.id,
                screeningTime = screening.screeningTime,
                screeningEndTime = screening.screeningEndTime,
                theater = TheaterResponse.from(screening.theater)
            )
        }
    }
}

data class TheaterResponse(
    val id: Long,
    val name: String
) {
    companion object {
        fun from(theater: Theater): TheaterResponse {
            return TheaterResponse(
                id = theater.id,
                name = theater.name
            )
        }
    }
}